testing/kokoro/version.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Determine if the given module should be tested in the current environment.
+//
+// If versions cannot be evaluated, we fail "successful" to ensure tests are
+// run by default.
+//
+// Usage:
+// go run . -module=../../run/hello-broken
+// This will exit with a non-zero status code if the module has a higher
+// go version than the runtime.
+
+// Command version checks for a module vs. environment go version mismatch
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/hashicorp/go-version"
+)
+
+// maxVersion is the most advanced go version in testing.
+const maxVersionStr="1.13"
+var maxVersion *version.Version
+
+func main() {
+	log.SetFlags(0)
+
+	modPathPtr := flag.String("module", ".", "Path to a go module")
+	flag.Parse()
+
+	v := strings.TrimLeft(runtime.Version(), "go")
+	envVersion, err := version.NewVersion(v)
+	if err != nil {
+		log.Printf("version.NewVersion: %v", err)
+	}
+	maxVersion, err := version.NewVersion(maxVersionStr)
+	if err != nil {
+		log.Printf("version.NewVersion: %v", err)
+	}
+
+	ok, err := valid(maxVersion, envVersion, *modPathPtr)
+	if err != nil {
+		log.Printf("compare: %v", err)
+	}
+	if !ok {
+		os.Exit(1)
+	}
+}
+
+// valid determines if the module should be tested.
+func valid(max *version.Version, v *version.Version, module string) (bool, error) {
+	if v.GreaterThan(max) {
+		log.Printf("always run tests for most advanced go version: go%s", maxVersionStr)
+		return true, nil
+	}
+
+	return compare(v, module)
```

Comment:
Maybe inline?

Reply:
I broke it out because the whole PR is intended as art for design discussion. Do you like the current approach or do you think we need to arrange this so there's a single `go run` to assemble the list of modules?

Reply:
I think this makes sense to me. "Run this binary to get the modules compatible with this version of Go." Right?

Reply:
Not exactly, right now it's "run this binary against every module you want to test to see if it should be tested". This means one run per module.

As an alternative, we could say "here's a list of modules, give me the list we should test". This means one two runs per test pipeline and moving more code into go.

Reply:
Decision: Polish up the current approach in this PR. Follow-up to consolidate on more go code with fewer go invocations.

Reply:
Ack